### ISSUE TITLE
Unable to load large files into pipeline

### ISSUE THREAD

**Issue by mscaldas2012** [CDCgov/data-exchange-hl7#388]
# TL;DR > Solution
~Remove content from hl7_recdeb_err queue payload.~

- **Truncate** `content` from hl7_recdeb_err queue payload.
- Intent: Tell the user why the message failed ==> Add why the message failed

# Details: 

Event Hub topic messages have a limit of 1MB in size.

So far any file size can be loaded to our Container  - and rightly so, since a file can be batched and should not cause an issue.

But, if bad files are uploaded, we are sending them to hl7-recdeb_err topic. Currently we are trying to include the file content as base64 encoded in a similar manner that we do for hl7-recdeb_ok.

But the hl7_recdeb_err does not need the content - only the metadata.

I suggest we do not include content for the error queue on receiver/debatcher - that can improve security and stability of pipeline (see error below when sending large files).

For traceability, the error queue will still have all the provenance information and they can track down the file on the storage and check its content to see what was sent. it just can't be handled by the pipeline.

The assumption is that a single HL7 message will NEVER reach 1MB or more and therefore our pipeliens are safe once we properly identify the payload as being HL7 and debatch it.

-- Error logs from Receiver/Debatcher function when uploading a 3Mb non-hl7 file:
```
2022-12-12T16:52:52.763 [Warning] [pool-2-thread-22] WARN com.azure.messaging.eventhubs.EventDataBatch - Size of the payload exceeded maximum message size: 1024 kb, errorContext[NAMESPACE: tf-eventhub-namespace-dev.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: hl7-recdeb-ok, REFERENCE_ID: 65C39B4C087D47CD087D47D463975C64_G11, LINK_CREDIT: 300]
2022-12-12T16:52:52.877 [Error] Executed 'Functions.receiverdebatcher001' (Failed, Id=0753acfd-aae6-4260-9005-425298b59252, Duration=759ms)Result: FailureException: AmqpException: Size of the payload exceeded maximum message size: 1024 kb, errorContext[NAMESPACE: tf-eventhub-namespace-dev.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: hl7-recdeb-ok, REFERENCE_ID: 65C39B4C087D47CD087D47D463975C64_G11, LINK_CREDIT: 300]Stack: java.lang.reflect.InvocationTargetExceptionat jdk.internal.reflect.GeneratedMethodAccessor3.invoke(Unknown Source)at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)at java.base/java.lang.reflect.Method.invoke(Method.java:566)at com.microsoft.azure.functions.worker.broker.JavaMethodInvokeInfo.invoke(JavaMethodInvokeInfo.java:22)at com.microsoft.azure.functions.worker.broker.EnhancedJavaMethodExecutorImpl.execute(EnhancedJavaMethodExecutorImpl.java:22)at com.microsoft.azure.functions.worker.chain.FunctionExecutionMiddleware.invoke(FunctionExecutionMiddleware.java:19)at com.microsoft.azure.functions.worker.chain.InvocationChain.doNext(InvocationChain.java:21)at com.microsoft.azure.functions.worker.broker.JavaFunctionBroker.invokeMethod(JavaFunctionBroker.java:80)at com.microsoft.azure.functions.worker.handler.InvocationRequestHandler.execute(InvocationRequestHandler.java:37)at com.microsoft.azure.functions.worker.handler.InvocationRequestHandler.execute(InvocationRequestHandler.java:12)at com.microsoft.azure.functions.worker.handler.MessageHandler.handle(MessageHandler.java:44)at com.microsoft.azure.functions.worker.JavaWorkerClient$StreamingMessagePeer.lambda$onNext$0(JavaWorkerClient.java:93)at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)at java.base/java.lang.Thread.run(Thread.java:829)Caused by: com.azure.core.amqp.exception.AmqpException: Size of the payload exceeded maximum message size: 1024 kb, errorContext[NAMESPACE: tf-eventhub-namespace-dev.servicebus.windows.net. ERROR CONTEXT: N/A, PATH: hl7-recdeb-ok, REFERENCE_ID: 65C39B4C087D47CD087D47D463975C64_G11, LINK_CREDIT: 300]at com.azure.messaging.eventhubs.EventDataBatch.tryAdd(EventDataBatch.java:126)at gov.cdc.dex.azure.EventHubSender.send(EventHubSender.kt:17)at gov.cdc.dex.hl7.receiver.Function.prepareAndSend(Function.kt:115)at gov.cdc.dex.hl7.receiver.Function.eventHubProcessor(Function.kt:103)... 17 more
```

The execution got flagged as Error - but this can still throw our counts out of wack, since that file will never be reported by the pipeline:

![image](https://user-images.githubusercontent.com/3239945/207107022-61407060-d4b1-4e0e-a837-c69257c0b129.png)

![image](https://user-images.githubusercontent.com/3239945/207107126-14d59101-abbc-4682-a3f1-14bb9ac1f0e3.png)

**Comment by rmharrison** [comment]
Resolution
- **Truncate** `content` from hl7_recdeb_err queue payload.
  * Don't remove the `content` key because you'll get "JSON Key not found" when running queries across all `err` queues
- Intent: Tell the user why the message failed
```
{
   "content": "Base64(MSH|^~\&|....)", <~~~ Truncate, don't remove key
   "summary": {
    	"current_status": "RECEIVED" <~~~ Option: Right now, the enum is `[RECIEVED, REJECTED]`, with no reason given [^1]
    },
        ...
       "processes": [
           {
            ...
            "status": "SUCCESS" <~~~ Option: Detail here [^2]
             ADDING "status_detail": "Content is not valid HL7v2" <~~~ Option: Add a status_detail 
           }
        ]
    },
...
}
```
[^1] https://github.com/CDCgov/data-exchange-hl7/blob/ca0df9341dc9e3fed1b4d465e6d055bb7c2c3efe/fns-hl7-pipeline/fn-receiver-debatcher/src/main/kotlin/gov/cdc/dex/hl7/receiver/Function.kt#L120
[^2] Although, it looks like "current_status" and "processes.status" are identical. https://github.com/CDCgov/data-exchange-hl7/blob/ca0df9341dc9e3fed1b4d465e6d055bb7c2c3efe/fns-hl7-pipeline/fn-receiver-debatcher/src/main/kotlin/gov/cdc/dex/hl7/receiver/Function.kt

Next steps
- @mscaldas2012 created Epic #393 to track long-term problem
- @marcia-schulman will implement solution

Consequences...
- Worse "ops ergonomics"
  * Now: The Ops engineer can see the full content
  * After change: The Ops engineer will have truncated content and a reason for rejection. They will have to refer back to the input message to see the full content.
- Punts on the problem of "externing" data in the EventHub events. Right now, the output of every function is appended to the events instead of being put in a separate location with a reference to that location put in the event.
